proposals/0046-astro-env.md

```diff
@@ -0,0 +1,594 @@
+import {
+  PUBLIC_PRODUCTION, // boolean | undefined
+  getSecret,
+} from "astro:env/server";
+
+const API_PORT = getSecret("API_PORT"); // number
+const UNKNOWN_VARIABLE = getSecret("UNKNOWN_VARIABLE"); // string | undefined
+```
+
+# Background & Motivation
+
+Env variables are an important part of any web application. You need to store sensitive data (think API secrets, tokens etc) without being leaked inside your git repo. But that's only the 1st part of the story. It's easy to leak this data by importing in the wrong place, eg. the frontend like [Resend in January 2024](https://resend.com/blog/incident-report-for-january-10-2024).
+
+Other JS frameworks (eg. [SvelteKit](https://kit.svelte.dev/docs/modules#$env-dynamic-private)) are handling env pretty well. However the env story is currently a bit tricky in Astro. According to the [docs](https://docs.astro.build/en/guides/environment-variables/), here is how env variables are currently handled:
+
+- Astro uses Vite’s built-in support for environment variables
+- Static variables (ie. replaced statically at build time) are accessible via `import.meta.env`
+- `import.meta.env` includes some [default variables](https://docs.astro.build/en/guides/environment-variables/#default-environment-variables) like `SSR`, `BASE_URL`...
+- A public variable key has to be prefixed by `PUBLIC_`
+- A non public variable accessed using `import.meta.env` on the client side will be `undefined` (value will be accessible server side)
+- Env variables can be loaded through `.env` (or `.env.production`, `.env.development`) and CLI
+- Any non built-in variable can be [manually typed](<[.env.production](https://docs.astro.build/en/guides/environment-variables/#intellisense-for-typescript)>)
+- Runtime variables should be access using `process.env`, or following the used runtime (eg. `Deno.env.get()` for the deno adapter)
+
+# Goals
+
+- Provide a fully type-safe experience for environment variables, without [manual type definitions](https://docs.astro.build/en/guides/environment-variables/#intellisense-for-typescript)
+- Reduce user confusion between inlined, static variables and dynamic, runtime variables
+- Allow adapters to specify how runtime env variables should be handled
+- Allow integrations to define environment variable constraints
+- Simple type casting and validation (strings/numbers/booleans, required/optional)
+
+# Non-Goals
+
+- Complex validation or type casting of env variables. We might want to enable this at some point, but there is likely a performance cost for this at runtime. We should punt on this if possible!
+- Future: allow adapters to customize build-time variable handling. Better runtime handling is the most important problem this proposal aims to solve, but if we find an API that enables this as well, that's great.
+
+# Detailed Design
+
+## Terminology
+
+- **Public variable**: variable replaced by its value at build time. Its key must be prefixed by `PUBLIC_`
```

Comment:
Do we need the `PUBLIC_` restriction here? The idea of forcing `PUBLIC_` in the past was, correct me if I'm wrong, to force you to think about the fact that it's possibly going to be exposed. But now, since you define the variables in the schema and have to use `access: 'public'`, I don't think the added prefix is adding any value here.

Reply:
The opposite (ie. `PUBLIC_` to be automatically public) was discussed [here](https://github.com/withastro/astro/pull/10974#discussion_r1595071201) and didn't really make sense to me. I initially wanted to keep the `PUBLIC_` prefix to match the behavior of `import.meta.env` (since it was supposed to be used originally) but I think you're right, it's not needed anymore. I'll submit a PR today

Reply:
Is `access` require? I think it should be. If not and public is the default, then this would allow someone to accidentally make something public and expose sensitive data.